src/main/java/org/apache/mesos/scheduler/DefaultScheduler.java

```diff
@@ -0,0 +1,311 @@
+        }
+
+    }
+
+    private void initializeRecoveryScheduler() {
+        logger.info("Initializing recovery scheduler");
+        RecoveryRequirementProvider recoveryRequirementProvider =
+                new DefaultRecoveryRequirementProvider(new DefaultOfferRequirementProvider());
+        LaunchConstrainer constrainer =
+                new TimedLaunchConstrainer(Duration.ofSeconds(DELAY_BETWEEN_DESTRUCTIVE_RECOVERIES_SEC));
+        recoveryScheduler = new DefaultRecoveryScheduler(
+                stateStore,
+                taskFailureListener,
+                recoveryRequirementProvider,
+                offerAccepter,
+                constrainer,
+                new TimedFailureMonitor(Duration.ofSeconds(PERMANENT_FAILURE_DELAY_SEC)),
+                recoveryStatusRef);
+    }
+
+    private void logOffers(List<Protos.Offer> offers) {
+        if (offers == null) {
+            return;
+        }
+
+        logger.info(String.format("Received %d offers:", offers.size()));
+        for (int i = 0; i < offers.size(); ++i) {
+            // Offer protobuffers are very long. print each as a single line:
+            logger.info(String.format("- Offer %d: %s", i + 1, TextFormat.shortDebugString(offers.get(i))));
+        }
+    }
+
+    private void declineOffers(SchedulerDriver driver, List<Protos.OfferID> acceptedOffers, List<Protos.Offer> offers) {
+        for (Protos.Offer offer : offers) {
+            Protos.OfferID offerId = offer.getId();
+            if (!acceptedOffers.contains(offerId)) {
+                logger.info("Declining offer: " + offerId.getValue());
+                driver.declineOffer(offerId);
+            }
+        }
+    }
+
+    private ResourceCleanerScheduler getCleanerScheduler() {
+        try {
+            ResourceCleaner cleaner = new ResourceCleaner(stateStore.getExpectedResources());
+            return new ResourceCleanerScheduler(cleaner, offerAccepter);
+        } catch (Exception ex) {
+            logger.error("Failed to construct ResourceCleaner", ex);
+            return null;
+        }
+    }
+
+    private List<Protos.Offer> filterAcceptedOffers(List<Protos.Offer> offers, List<Protos.OfferID> acceptedOfferIds) {
+        List<Protos.Offer> filteredOffers = new ArrayList<>();
+
+        for (Protos.Offer offer : offers) {
+            if (!offerAccepted(offer, acceptedOfferIds)) {
+                filteredOffers.add(offer);
+            }
+        }
+
+        return filteredOffers;
+    }
+
+    private boolean offerAccepted(Protos.Offer offer, List<Protos.OfferID> acceptedOfferIds) {
+        for (Protos.OfferID acceptedOfferId: acceptedOfferIds) {
+            if (acceptedOfferId.equals(offer.getId())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @SuppressWarnings({"DM_EXIT"})
+    private void hardExit(SchedulerErrorCode errorCode) {
+        System.exit(errorCode.ordinal());
```

Comment:
This is logged by all reasonably implemented Executors including the CmdExecutor used by Marathon.